### ISSUE TITLE
Multiple workspace per Brix -> SubWorkspaces

### ISSUE THREAD

**Issue by kbachl** [brix-cms/brix-cms#196]
Currently brix uses 1 workspace for 1 site; This limits some advanced features like the new Brix CMS Plugin, as putting all data into 1 workspace seems not a good idea as the data can grow too far out;

Therefore we need to bring some changes to brix to allow 1 Site to have multiple workspaces;

The basic idea here is to use the dash (.) to seperate the workspaces, having the current SiteWorkspace the "mother" of all of it, so the lineup would be:

SiteWorkspace
SiteWorkspace.SubWorkspaceA
SiteWorkspace.SubWorkspaceB
SiteWorkspace.SubWorkspaceC ...

**To allow working with this we need to enhance at some points:**

1. the createWorkspace function must check for dashes (.) and disallow their usage
2. a new function createSubWorkspace(String mainWorkspaceName, String subWorkspaceName) is needed
3. the workspace-switcher admin-backend needs to be refactored to have its level limited to the "SiteWorkspace" while adding a second switcher to eventually allow to switch from main-SiteWorkspace through the SiteWorkspace.SubWorkspaces
4. this has to be done to the frontend as well
5. the export/ import/ snapshot functions have to be advanced in a way that they allow to include all SiteWorkspace.SubWorkspaces in their operations (has to be optional)

At a later time it can even be researched if it maybe suitable to refactor the current SitePlugin (the main brix module) into a SubWorkspace itself haveing the SiteWorkspace solely for the job of delivering Brix-Meta and Brix-Config data about the plugins and Subworkspaces - **however, since this would break all brix-apps it would need to have a converter who will convert old styled Workspaces to the new format!**

**Comment by dsimko** [comment]
I don't think that this is needed. IMHO Brix was designed in order to plugins can create their workspaces. e.g. look at https://github.com/brix-cms/brix-cms/blob/master/brix-core/src/main/java/org/brixcms/Plugin.java#L96  I don't see any reason why these workspaces should be so tightly coupled with site workspace.

**Comment by kbachl** [comment]
Well, I'm sorry but I think having some limits like disallowing "." in workspace name etc. is really needed - I dont want brix to end up the same crap like wordpress, joomla etc. where a plugin can do any sh**t a dev wants to. I see brix as a CMS for professional needs, not the typical hipster crap that breaks as soon as 1 year went by...

Plus the adding to the switcher would allow of course to hide workspaces that dont want to be shown there; Plus, going later on to a central meta workspace and all plugins in their own or the siteworkspace would allow to make use of dynamic plugin/ tile deplyoing (java 9 -> jigsaw);

this isn't for version 8, but we should see where we want brix to grow to....

**Comment by dsimko** [comment]
> Well, I'm sorry but I think having some limits like disallowing "." in workspace name etc. is really needed

I am still not sure whether it's needed. And if so then dot "." is probably not the best separator because site workspace has usually name created from domain or subdomain.

**Comment by kbachl** [comment]
Dan, I dont care the sign - thats childs play! Make it "#" or any other non A-Z sign; But we need to think for the future; Imposing a limit as early as a introducing a 2nd workspace from brix itself is necessary - as its easy to lift later, while introducing it later is always the danger of breaking things later (on the userside);

I know that there are subworkspaces in the brix code, but this was one of the parts that never saw real action and isn't really done nicely. Brix started as a way to power many webpages for a wine-seller startup in the US - but the company abandoned it long ago and so did Igor and Matej Knoop, the original authors. Heck, even the JCR node definitions didnt work together with modeshape in the beginning as they were violating some part of the JCR spec...

**Comment by kbachl** [comment]
PS: this task is for 8.0.0 final - so you will get a working 8.0.0.beta1 and more betas for you to start early enough....

**Comment by dsimko** [comment]
> PS: this task is for 8.0.0 final

ok, we will consider it later. 

One more point which I would like to mention is that content-plugin should be usable even without site-plugin. One can use only admin part of content-plugin and display this data on a regular wicket page.

**Comment by kbachl** [comment]
>One more point which I would like to mention is that content-plugin should be usable even without site-plugin. One can use only admin part of content-plugin and display this data on a regular wicket page.

I thought about that and that's exactly the reason why I want to go to a Brix-Meta main workspace in future, so even a SitePlugin is optional... (To Discuss for the Future);
